core/dataset.cpp

```diff
@@ -310,6 +310,44 @@ void Dataset::setData(const std::string &name, Variable data) {
     rebuildDims();
 }
 
+/// Set (insert or replace) data from a DataArray with a given name, avoiding
+/// copies where possible by using std::move.
+///
+void Dataset::setDataMove(const std::string &name, DataArray data) {
+  // Get the Dataset holder
+  auto dataset = DataArray::to_dataset(std::move(data));
+
+  for (const auto &[dim, coord] : dataset.m_coords) {
+    if (const auto it = m_coords.find(dim); it != m_coords.end())
+      expect::equals(coord, it->second);
+    else
+      setCoord(dim, std::move(coord));
+  }
+  for (const auto &[nm, labs] : dataset.m_labels) {
+    if (const auto it = m_labels.find(std::string(nm)); it != m_labels.end())
+      expect::equals(labs, it->second);
+    else
+      setLabels(std::string(nm), std::move(labs));
+  }
+
+  for (const auto &[nm, mask] : dataset.m_masks)
+    setMask(std::string(nm), std::move(mask));
```

Comment:
Unrelated (was done in the existing `setData` as well): I wonder whether just setting the mask is ok, shouldn't we check if it exists first? `setMask` will just replace the existing one.